packages/auto-router/src/lib/savings-vault-router.ts

```diff
@@ -0,0 +1,240 @@
+import { Erc20, StandardPermitArguments } from '@phuture/erc-20';
+import { SavingsVault } from '@phuture/savings-vault';
+import { Address } from '@phuture/types';
+import { BigNumber, BigNumberish, constants } from 'ethers';
+import { TransactionResponse } from '@ethersproject/abstract-provider';
+import { InsufficientAllowanceError, PhutureError } from '@phuture/errors';
+import { Router } from './interfaces';
+
+/** ### SavingsVaultRouter class */
+export class SavingsVaultRouter implements Router {
+	/**
+	 * ### Select Buy
+	 *
+	 * @param savingsVault Contract which implements the SavingsVault interface
+	 * @param amountInInputToken Amount in input token
+	 * @param inputToken Erc20 or Erc20Permit interface of input token
+	 *
+	 * @returns isMint True if minting selected.
+	 * @returns target Address of the target contract
+	 * @returns output Amount of Index shares
+	 * @returns expectedAllowance Allowance for the input token
+	 */
+	async selectBuy(
+		savingsVault: SavingsVault,
+		amountInInputToken: BigNumberish,
+		inputToken?: Erc20
+	): Promise<{
+		isMint: boolean;
+		target: Address;
+		outputAmount: BigNumber;
+		expectedAllowance?: BigNumber;
+	}> {
+		let target = constants.AddressZero;
+
+		let expectedAllowance: BigNumber | undefined;
+		if (inputToken) {
+			if (
+				inputToken.address.toLowerCase() !==
+				(await savingsVault.contract.asset()).toLowerCase()
+			) {
+				throw new PhutureError({
+					status: 400,
+					message:
+						'Input token is not the underlying asset token of the SavingsVault',
+				});
+			}
+			try {
+				await inputToken.checkAllowance(
+					savingsVault.address,
+					amountInInputToken
+				);
+			} catch (error) {
+				if (error instanceof InsufficientAllowanceError) {
+					target = savingsVault.address;
+					expectedAllowance = error.expectedAllowance;
```

Comment:
you can return an object from function here

Reply:
Not sure what you mean